aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -0,0 +1,309 @@
+_check_last_cidx_is_nnz(const index_t& cidx, const index_t& nnz) {
+  const bool invariant = cidx == nnz;
+  if (cdim_name == CDimName::CRow) {
+    _assert(invariant, "`crow_indices[..., -1] == nnz` is not satisfied.");
+  }
+  else {
+    _assert(invariant, "`ccol_indices[..., -1] == nnz` is not satisfied.");
+  }
+}
+
+// Invariant 5.3
+// 0 <= compressed_indices[..., 1:] - compressed_indices[..., :-1] <= plain_dim.
+template <CDimName cdim_name, typename index_t>
+INVARIANT_CHECK_FUNC_API
+_check_cidx_nondecreasing_locally_bounded_sequence(
+    const index_t& cidx,
+    const index_t& cidx_next,
+    const index_t& zero,
+    const index_t& dim) {
+  const auto s_cidx = cidx_next - cidx;
+  const bool invariant = zero <= s_cidx && s_cidx <= dim;
+  if (cdim_name == CDimName::CRow) {
+    _assert(invariant,
+        "`0 <= crow_indices[..., 1:] - crow_indices[..., :-1] <= ncols` is not satisfied.");
+  }
+  else {
+    _assert(invariant,
+        "`0 <= ccol_indices[..., 1:] - ccol_indices[..., :-1] <= nrows` is not satisfied.");
+  }
+}
+
+// Invariants 5.4 and 5.5
+// 0 <= plain_index < plain_dim.
+template <CDimName cdim_name, typename index_t>
+INVARIANT_CHECK_FUNC_API
+_check_idx_bounds(
+    const index_t& idx,
+    const index_t& zero,
+    const index_t& dim) {
+  const bool invariant = zero <= idx && idx < dim;
+  if (cdim_name == CDimName::CRow) {
+    _assert(invariant, "`0 <= col_indices < ncols` is not satisfied.");
+  }
+  else {
+    _assert(invariant, "`0 <= row_indices < nrows` is not satisfied.");
+  }
+}
+
+// Invariant 5.6
+// plain_indices[..., compressed_indices[..., i - 1]:compressed_indices[..., i]]
+// for all i = 1, ..., compressed_dim
+// are sorted and distinct along the last dimension values.
+template <CDimName cdim_name, typename index_t>
+INVARIANT_CHECK_FUNC_API
+_check_idx_sorted_distinct_vals_slices_with_cidx(
+    const index_t* RESTRICT ptr_idx_batch,
+    const index_t cidx,
+    const index_t cidx_next) {
+  // Note that ptr_idx_batch = &idx[batch_idx] and is contiguous.
+  const auto* RESTRICT slice_begin = ptr_idx_batch + cidx;
+  const auto* RESTRICT slice_end = ptr_idx_batch + cidx_next;
+  for (auto* RESTRICT curr = slice_begin + 1; curr < slice_end; ++curr) {
+    const auto invariant = *(curr - 1) < *curr;
+    if (cdim_name == CDimName::CRow) {
+      _assert(invariant, "`col_indices[..., crow_indices[..., i - 1]:crow_indices[..., i]] "
+                         "for all i = 1, ..., nrows "
+                         "are sorted and distinct along the last dimension values` "
+                         "is not satisfied.");
+    }
+    else {
+      _assert(invariant, "`row_indices[..., ccol_indices[..., i - 1]:ccol_indices[..., i]] "
+                         "for all i = 1, ..., ncols "
+                         "are sorted and distinct along the last dimension values` "
+                         "is not satisfied.");
+    }
+  }
+}
+
+static inline int64_t numel(IntArrayRef sizes) {
```

Comment:
By the usage, this function counts the number of batches (or more generally, the number of indices), not the number of elements, and with a property `sizes==()` maps to `1`.
Rename `numel` -> `indexCount` or similar?